src/fsharp/NameResolution.fs

```diff
@@ -770,57 +788,376 @@ let AddUnionCases2 bulkAddMode (eUnqualifiedItems: UnqualifiedItems) (ucrefs: Un
             let item = Item.UnionCase(GeneralizeUnionCaseRef ucref, false)
             acc.Add (ucref.CaseName, item))
 
-let AddStaticContentOfTyconRefToNameEnv (g:TcGlobals) (amap: Import.ImportMap) ad m (nenv: NameResolutionEnv) (tcref:TyconRef) =
-    // If OpenStaticClasses is not enabled then don't do this
-    if amap.g.langVersion.SupportsFeature LanguageFeature.OpenStaticClasses then
-        let ty = generalizedTyconRef tcref
-        let infoReader = InfoReader(g,amap)
-        let items =
-            [| let methGroups = 
-                   AllMethInfosOfTypeInScope ResultCollectionSettings.AllResults infoReader nenv None ad PreferOverrides m ty
-                   |> List.groupBy (fun m -> m.LogicalName)
-
-               for (methName, methGroup) in methGroups do
-                   let methGroup = methGroup |> List.filter (fun m -> not m.IsInstance && not m.IsClassConstructor)
-                   if not methGroup.IsEmpty then
-                       yield KeyValuePair(methName, Item.MethodGroup(methName, methGroup, None)) 
-           
-               let propInfos = 
-                   AllPropInfosOfTypeInScope ResultCollectionSettings.AllResults infoReader nenv None ad PreferOverrides m ty
-                   |> List.groupBy (fun m -> m.PropertyName)
-               
-               for (propName, propInfos) in propInfos do
-                   let propInfos = propInfos |> List.filter (fun m -> m.IsStatic)
-                   for propInfo in propInfos do 
-                       yield KeyValuePair(propName , Item.Property(propName,[propInfo]))
-            
-               let fields =
-                  infoReader.GetILFieldInfosOfType(None, ad, m, ty)
-                  |> List.groupBy (fun f -> f.FieldName)
-
-               for (fieldName, fieldInfos) in fields do
-                   let fieldInfos = fieldInfos |> List.filter (fun fi -> fi.IsStatic)
-                   for fieldInfo in fieldInfos do
-                       yield KeyValuePair(fieldName, Item.ILField(fieldInfo))
-             |]
-
-        { nenv with eUnqualifiedItems = nenv.eUnqualifiedItems.AddAndMarkAsCollapsible items }
-    else
-        nenv
+//-------------------------------------------------------------------------
+// TypeNameResolutionInfo
+//-------------------------------------------------------------------------
+
+/// Indicates whether we are resolving type names to type definitions or to constructor methods.
+type TypeNameResolutionFlag =
+    | ResolveTypeNamesToCtors
+    | ResolveTypeNamesToTypeRefs
+
+[<RequireQualifiedAccess>]
+[<NoEquality; NoComparison>]
+/// Represents information about the generic argument count of a type name when resolving it.
+///
+/// In some situations we resolve "List" to any type definition with that name regardless of the number
+/// of generic arguments. In others, we know precisely how many generic arguments are needed.
```

Comment:
These comments are not mine nor is a lot of code here. If you look closely, I had to move these types and functions closer to the top of the file in order to access them in the functions that I needed. I would rather not try to refactor them just because they show up as an addition in the diff.

Reply:
I agree, minimum necessary delta is a decent policy here.